### ISSUE TITLE
Violated stack walking assertion when using debug ILCompiler build

### ISSUE THREAD

**Issue by AntonLapounov** [dotnet/runtimelab#988]
Create a console app containing just this line:
```C#
throw new System.Exception();
```
and compile it in release mode with a **debug** build of the ILCompiler package:
```cmd
dotnet publish -r win-x64 -c Release
```
Run it and notice the following violated assertion reported:
```
--------------------------------------------------
Debug Assertion Violation

Expression: 'pCurExInfo->m_idxCurClause != MaxTryRegionIdx'

File: src\coreclr\nativeaot\Runtime\StackFrameIterator.cpp, Line: 1850
--------------------------------------------------
```
with the following call stack:
```
Repro!Assert
Repro!RhpSfiNext
Repro!S_P_CoreLib_System_Runtime_RuntimeExports__RhpCalculateStackTraceWorker
Repro!RhpGetCurrentThreadStackTrace
Repro!S_P_CoreLib_System_Runtime_RuntimeExports__RhGetCurrentThreadStackTrace
Repro!S_P_CoreLib_System_Diagnostics_StackTrace__InitializeForCurrentThread
Repro!S_P_CoreLib_System_Diagnostics_StackTrace___ctor_3
Repro!S_P_CoreLib_System_Diagnostics_DebugProvider__Fail
Repro!S_P_CoreLib_System_Diagnostics_Debug__Fail_0
Repro!S_P_CoreLib_System_Diagnostics_Debug__Assert_1
Repro!S_P_CoreLib_System_Diagnostics_Debug__Assert
Repro!S_P_TypeLoader_Internal_Runtime_TypeLoader_TypeLoaderEnvironment__TryGetStaticClassConstructionContext
Repro!S_P_Reflection_Execution_Internal_Reflection_Execution_ReflectionExecutionDomainCallbacksImplementation__TryGetStaticClassConstructionContext
Repro!S_P_CoreLib_System_Runtime_CompilerServices_RuntimeHelpers__RunClassConstructor
Repro!S_P_CoreLib_System_SR__InternalGetResourceString
Repro!S_P_CoreLib_System_SR__GetResourceString_0
Repro!S_P_CoreLib_System_Exception__get_Message
Repro!S_P_CoreLib_System_Exception__AppendExceptionStackFrame
Repro!S_P_CoreLib_System_Runtime_EH__AppendExceptionStackFrameViaClasslib
Repro!S_P_CoreLib_System_Runtime_EH__UpdateStackTrace
Repro!S_P_CoreLib_System_Runtime_EH__DispatchEx
Repro!S_P_CoreLib_System_Runtime_EH__RhThrowEx
Repro!RhpThrowEx
Repro!Repro__Program_____Main__
Repro!Repro__Module___StartupCodeMain
Repro!wmain
Repro!invoke_main
Repro!__scrt_common_main_seh
Repro!__scrt_common_main
Repro!wmainCRTStartup
KERNEL32!BaseThreadInitThunk
ntdll!RtlUserThreadStart
```
See #987 for instructions on creating a debug ILCompiler package.

**Comment by MichalStrehovsky** [comment]
Here's a repro that doesn't depend on hitting an assertion failure during dispatch:

```csharp
try
{
    throw new Exception("");
}
catch when (Filter())
{
}

static bool Filter()
{
    Console.WriteLine(new StackTrace(0, true));
    return true;
}
```

Fails with:

```
>	reproNative.exe!Assert(const char * expr, const char * file, unsigned int line_num, const char * message) Line 49	C++
 	reproNative.exe!RhpSfiNext(StackFrameIterator * pThis, unsigned int * puExCollideClauseIdx, unsigned char * pfUnwoundReversePInvoke) Line 1851	C++
 	reproNative.exe!S_P_CoreLib_System_Runtime_RuntimeExports__RhpCalculateStackTraceWorker() Line 340	Unknown
 	reproNative.exe!RhpGetCurrentThreadStackTrace(void * pOutputBuffer, unsigned int outputBufferLength) Line 394	C++
 	reproNative.exe!S_P_CoreLib_System_Runtime_RuntimeExports__RhGetCurrentThreadStackTrace() Line 310	Unknown
 	reproNative.exe!S_P_CoreLib_System_Diagnostics_StackTrace__InitializeForCurrentThread() Line 33	Unknown
 	reproNative.exe!S_P_CoreLib_System_Diagnostics_StackTrace___ctor_3() Line 70	Unknown
 	reproNative.exe!repro__Program______Main___g__Filter_0_0() Line 18	Unknown
 	reproNative.exe!repro__Program_____Main__() Line 12	Unknown
 	reproNative.exe!RhpCallFilterFunclet()	Unknown
 	reproNative.exe!S_P_CoreLib_System_Runtime_EH__FindFirstPassHandler() Line 858	Unknown
 	reproNative.exe!S_P_CoreLib_System_Runtime_EH__DispatchEx() Line 687	Unknown
 	reproNative.exe!S_P_CoreLib_System_Runtime_EH__RhThrowEx() Line 618	Unknown
 	reproNative.exe!RhpThrowEx()	Unknown
 	reproNative.exe!repro__Program_____Main__() Line 10	Unknown
 	reproNative.exe!repro__Module___StartupCodeMain()	Unknown

```

**Comment by jkotas** [comment]
Fixed by #1153
